sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/utils/SchemaUtil.scala

```diff
@@ -81,6 +84,112 @@ object SchemaUtil {
+      partitionId: Int): Iterator[InternalRow] = {
+    val groupingKeySchema = SchemaUtil.getSchemaAsDataType(
+      compositeKeySchema, "key"
+    ).asInstanceOf[StructType]
+    val userKeySchema = SchemaUtil.getSchemaAsDataType(
+      compositeKeySchema, "userKey"
+    ).asInstanceOf[StructType]
+
+    def appendKVPairToMap(
+        curMap: mutable.Map[Any, Any],
+        stateRowPair: UnsafeRowPair): Unit = {
+      curMap += (
+        stateRowPair.key.get(1, userKeySchema)
+          .asInstanceOf[UnsafeRow].copy() ->
+          stateRowPair.value.copy()
+        )
+    }
+
+    def updateDataRow(
+        groupingKey: Any,
+        curMap: mutable.Map[Any, Any]): GenericInternalRow = {
+      val row = new GenericInternalRow(3)
+      val mapData = new ArrayBasedMapData(
+        ArrayData.toArrayData(curMap.keys.toArray),
+        ArrayData.toArrayData(curMap.values.toArray)
+      )
+      row.update(0, groupingKey)
+      row.update(1, mapData)
+      row.update(2, partitionId)
+      row
+    }
+
+    // All of the rows with the same grouping key were co-located and were
+    // grouped together consecutively.
+    new Iterator[InternalRow] {
+      var curGroupingKey: UnsafeRow = _
+      var curStateRowPair: UnsafeRowPair = _
+      val curMap = mutable.Map.empty[Any, Any]
+
+      override def hasNext: Boolean =
+        stateRows.hasNext || !curMap.isEmpty
+
+      override def next(): InternalRow = {
+        var keepTraverse = true
+        while (stateRows.hasNext && keepTraverse) {
+          curStateRowPair = stateRows.next()
+          if (curGroupingKey == null) {
+            // First time in the iterator
+            // Need to make a copy because we need to keep the
+            // value across function calls
+            curGroupingKey = curStateRowPair.key
+              .get(0, groupingKeySchema).asInstanceOf[UnsafeRow].copy()
+            appendKVPairToMap(curMap, curStateRowPair)
+          } else {
+            val curPairGroupingKey =
+              curStateRowPair.key.get(0, groupingKeySchema)
+            if (curPairGroupingKey == curGroupingKey) {
+              appendKVPairToMap(curMap, curStateRowPair)
+            } else {
+              // find a different grouping key, exit loop and return a row
+              keepTraverse = false
+            }
+          }
+        }
+        if (!keepTraverse) {
+          // found a different grouping key
+          val row = updateDataRow(curGroupingKey, curMap)
+          // update vars
+          curGroupingKey =
+            curStateRowPair.key.get(0, groupingKeySchema)
+              .asInstanceOf[UnsafeRow].copy()
+          // empty the map, append current row
+          curMap.clear()
+          appendKVPairToMap(curMap, curStateRowPair)
+          // return map value of previous grouping key
+          row
+        } else {
+          // reach the end of the state rows
+          if (curMap.isEmpty) null.asInstanceOf[InternalRow]
```

Comment:
This shouldn't happen if caller checks hasNext before, right? There is no case where null as InternalRow is expected.

Let's follow Iterator's contract. We should throw NoSuchElementException, or if we want to track this separately, use Spark's internal exception.

Reply:
Sure, that makes a lot of sense. I modified this to be throw `NoSuchElementException` the same as other iterators.